### PR TITLE
feat: EXCEPT in SELECT

### LINKED ISSUE #9092: feat: SELECT * EXCEPT [columnA,columnB,...]
It is a SQL grammar supported by Snowflake 
It can improve the usability of GroupTopN pattern https://www.risingwave.dev/docs/current/sql-pattern-topn/
Still need discussion. 
```SQL
SELECT * EXCEPT(rank)
  FROM (
    SELECT [column_list], 
      function_name() OVER ([PARTITION BY col1[, col2...]] 
        ORDER BY col1 [ ASC | DESC ][, col2 [ ASC | DESC ]...]) AS rank 
    FROM table_name)
WHERE rank <= N [AND conditions];
```

### PR DESCRIPTION
**Pull request by wugouzi**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
Add support to following syntax:
```
SELECT * EXCEPT (column1, column2, ...)
...
```

close #9092.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

## Documentation

- [x] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Comment by lmatz** [comment]
Can we add more tests, e.g.:

1. `except` in an inner subquery
2. `except` in the outer query when there is an inner subquery
3.  a combination of 1 + 2

**Review by st1page — COMMENTED** [review]
1. we can not bind it in the query level, we should treat it as a modifier on the `*` or WildCard.
```SQL
dev=# select * from t;
 a | b | c 
---+---+---
 2 | 1 | 3
(1 row)

dev=# select *, * from t;
 a | b | c | a | b | c 
---+---+---+---+---+---
 2 | 1 | 3 | 2 | 1 | 3
(1 row)
```
2. we need to support multiple value in the except. `SELECT * EXCEPT (i, j, k);`

In conclusion, we might need to add `SelectItem::ExceptWithExcept(Vec<Expr>)` even use it to replace the original WildCard

File: src/sqlparser/tests/testdata/select.yaml [hunk 1]
```diff
@@ -37,6 +37,8 @@
   formatted_sql: SELECT id, fname, lname FROM customer WHERE salary <> 'Not Provided' AND salary <> ''
 - input: SELECT id FROM customer WHERE NOT salary = ''
   formatted_sql: SELECT id FROM customer WHERE NOT salary = ''
+- input: SELECT * EXCEPT v1 FROM foo
+  formatted_sql: SELECT EXCEPT v1 FROM foo
```
Comment (xxchan) [inline src/sqlparser/tests/testdata/select.yaml:41]: Is this valid?

**Comment by xxchan** [comment]
> except in the outer query when there is an inner subquery

e.g., the TopN pattern

---

BTW, is the syntax decided?
- Whether to use `EXCEPT` (clickhouse, spanner) or `EXCLUDE` (snowflakes, duckdb)
- Whether `()` is required (all 4 DBs above seem to require that). BTW, multiple columns are not tested and things like `select a,b except` are not tested.

**Comment by kwannoel** [comment]
Hi could you add some sqlsmith fuzzing tests? Feel free to ask me if any questions.

You may check out the developer docs here: https://github.com/risingwavelabs/risingwave/blob/main/src/tests/sqlsmith/develop.md

File: src/sqlparser/tests/testdata/select.yaml (hunk 1 above)
Comment (wugouzi) [reply src/sqlparser/tests/testdata/select.yaml:41]: This is valid in current implementation and there are many things need to be fixed and there will be possible change later.

**Comment by xxchan** [comment]
You shared an `EXCLUDE` doc 😄

image

**Comment by st1page** [comment]
> You shared an `EXCLUDE` doc smile
> 
> image

yes and just to say that all the grammer need `()`

File: src/frontend/planner_test/tests/testdata/output/basic_query.yaml [hunk 2]
```diff
@@ -42,6 +42,15 @@
     StreamMaterialize { columns: [v1, t._row_id(hidden)], stream_key: [t._row_id], pk_columns: [t._row_id], pk_conflict: "NoCheck" }
     └─StreamFilter { predicate: (t.v1 < 1:Int32) }
       └─StreamTableScan { table: t, columns: [t.v1, t._row_id], pk: [t._row_id], dist: UpstreamHashShard(t._row_id) }
+- sql: |
+    create table t (v1 int, v2 int, v3 int);
+    select * except (v1, v2) from t;
```
Comment (kwannoel) [inline src/frontend/planner_test/tests/testdata/output/basic_query.yaml:47]: Can we add more tests which explicitly select `except` columns?
Something like:
```sql
select v3 from (select * except (v1, v2) from t
```

Can we also add tests which have selected the excepted row, and re-select already included rows?
```
select * except (v1, v2), v1 from t; -- v1 excluded from except.
select * except (v1, v2), v3 from t; -- v3 included from except.
```

File: src/sqlparser/src/ast/mod.rs [hunk 3]
```diff
@@ -1919,6 +1919,7 @@ pub enum FunctionArgExpr {
     QualifiedWildcard(ObjectName),
     /// An unqualified `*`
     Wildcard,
+    Except(Vec<Expr>),
```
Comment (st1page) [inline src/sqlparser/src/ast/mod.rs:1922]: rename it to `WildcardWithExcept`. a little confusing here.

File: src/frontend/src/binder/select.rs [hunk 4]
```diff
@@ -259,40 +259,63 @@ impl Binder {
...
+                            select_list.extend(exprs);
+                            aliases.extend(names);
+                        }
+                        None => {
+                            if self.context.range_of.is_empty() {
+                                return Err(ErrorCode::BindError(
+                                    "SELECT * with no tables specified is not valid".into(),
+                                )
+                                .into());
+                            }
```
Comment (st1page) [inline src/frontend/src/binder/select.rs:288]: extract it out of the match?

File: src/frontend/src/binder/select.rs [hunk 5]
```diff
@@ -259,40 +259,63 @@ impl Binder {
...
+                            let (exprs, names) = Self::iter_bound_columns(
+                                self.context.columns[..].iter().filter(|c| {
+                                    !c.is_hidden
+                                        && !self
+                                            .context
+                                            .column_group_context
+                                            .mapping
+                                            .contains_key(&c.index)
+                                }),
+                            );
```
Comment (st1page) [inline src/frontend/src/binder/select.rs:306]: I think some common logic about the column_groups is lost

**Review by st1page — APPROVED** [review]
LGTM

**Review by xxchan — COMMENTED** [review]
There aren't any tests for syntax errors 😟

**Comment by wugouzi** [comment]
I will add some.

File: src/sqlparser/tests/testdata/select.yaml [hunk 6]
```diff
@@ -29,14 +29,21 @@
...
 - input: SELECT id, fname, lname FROM customer WHERE salary <> 'Not Provided' AND salary <> ''
   formatted_sql: SELECT id, fname, lname FROM customer WHERE salary <> 'Not Provided' AND salary <> ''
 - input: SELECT id FROM customer WHERE NOT salary = ''
   formatted_sql: SELECT id FROM customer WHERE NOT salary = ''
+- input: SELECT * EXCEPT (v1,v2) FROM foo
+  formatted_sql: SELECT * EXCEPT (v1, v2) FROM foo
+  formatted_ast: 'Query(Query { with: None, body: Select(Select { distinct: All, projection: [WildcardOrWithExcept(Some([Identifier(Ident { value: "v1", quote_style: None }), Identifier(Ident { value: "v2", quote_style: None })]))], from: [TableWithJoins { relation: Table { name: ObjectName([Ident { value: "foo", quote_style: None }]), alias: None, for_system_time_as_of_proctime: false }, joins: [] }], lateral_views: [], selection: None, group_by: [], having: None }), order_by: [], limit: None, offset: None, fetch: None })'
+- input: SELECT v3 EXCEPT (v1, v2) FROM foo
+  error_msg: |-
```
Comment (wugouzi) [inline src/sqlparser/tests/testdata/select.yaml:44]: I think this is a test for syntax. But it is not sufficient indeed. I will add some.

**Pull request merged**